spatial_cell_gpu.cpp

```diff
@@ -0,0 +1,1170 @@
+         dev_velocity_block_with_content_map = velocity_block_with_content_map->upload<true>(stream);
+      }
+      if (vbwncl_sizePower < HashmapReqSize) {
+         vbwncl_sizePower = HashmapReqSize;
+         ::delete velocity_block_with_no_content_map;
+         void *buf = malloc(sizeof(Hashinator::Hashmap<vmesh::GlobalID,vmesh::LocalID>));
+         velocity_block_with_no_content_map = ::new (buf) Hashinator::Hashmap<vmesh::GlobalID,vmesh::LocalID>(vbwncl_sizePower);
+         dev_velocity_block_with_no_content_map = velocity_block_with_no_content_map->upload<true>(stream);
+      }
+      // These lists are also used in acceleration, where sometimes, very many blocks may be added.
+      // Thus, this one list needs to have larger capacity than the others..
+      if (list_with_replace_new_capacity < reserveSize * acc_reserve_multiplier) {
+         list_with_replace_new->reserve(newReserve * acc_reserve_multiplier,true);
+         list_with_replace_new_capacity = newReserve * acc_reserve_multiplier;
+         dev_list_with_replace_new = list_with_replace_new->upload<true>(stream);
+      }
+      if (list_delete_capacity < reserveSize) {
+         list_delete->reserve(newReserve,true);
+         list_delete_capacity = newReserve;
+         dev_list_delete = list_delete->upload<true>(stream);
+      }
+      if (list_to_replace_capacity < reserveSize) {
+         list_to_replace->reserve(newReserve,true);
+         list_to_replace_capacity = newReserve;
+         dev_list_to_replace = list_to_replace->upload<true>(stream);
+      }
+      if (list_with_replace_old_capacity < reserveSize) {
+         list_with_replace_old->reserve(newReserve,true);
+         list_with_replace_old_capacity = newReserve;
+         dev_list_with_replace_old = list_with_replace_old->upload<true>(stream);
+      }
+   }
+
+   /** Adds "important" and removes "unimportant" velocity blocks
+    * to/from this cell.
+    *
+    * velocity_block_with_content_list needs to be up to date in local and remote cells.
+    * velocity_block_with_no_content_list needs to be up to date in local cells.
+    *
+    * update_velocity_block_with_content_lists() should have
+    * been called with the current distribution function values, and then the contetn list transferred.
+    *
+    * Removes all velocity blocks from this spatial cell which don't
+    * have content and don't have spatial or velocity neighbors with
+    * content.  Adds neighbors for all velocity blocks which do have
+    * content (including spatial neighbors).  All cells in
+    * spatial_neighbors are assumed to be neighbors of this cell.
+    *
+    * This function is thread-safe when called for different cells
+    * per thread. We need the block_has_content vector from
+    * neighbouring cells, but these are not written to here. We only
+    * modify local cell.*/
+
+   void SpatialCell::adjust_velocity_blocks(const uint popID, bool doDeleteEmptyBlocks) {
```

Comment:
Given that the batched version of block adjustment now exists and seems to be the primary workhorse, does it make sense to keep this second, redundant version around?
Obviously there are usecases for it, but is there any way to reduce the code duplication?

Reply:
We use this version e.g. after creating new template cells for time-varying boundaries, and using the batch approach for a single cell is overkill.